Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,227 @@
+- Stay backwards compatible to the current mechanism by default
+
+## Non-goals
+
+- Implement mechanisms to avoid configuration errors by the user
+- Support mixed environments (kubernetes and non-kubernetes targets are scraped)
+
+## Algorithm
+
+In order to do calculate a stable assignment, following parameters are required:
+
+1. `num_shards`: The number of prometheus shards
+2. `shard_index`: A number of the range `[0..num_shards[` identifying a single prometheus 
+   instance
+3. `zones`: A list of the zones to be scraped
+4. `zone_label`: A label denoting the zone of a target
+5. `address`: The content of the `__address__` label
+
+It has to be noted that `zone_label` is supposed to contain a value from the
+`zones` list.
+The `num_shards` value is referring to the currently available `.spec.shards`
+from the [Prometheus custom resource definition](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1.Prometheus).
+
+Given these values, a target is to be scraped by a given prometheus instance
+by using the following algorithm:
+
+```go
+assert(num_shards >= len(zones))     # Error: zone(s) will not get scraped
+assert(num_shards % len(zones) == 0) # Warning: multi-scraping of instances
+
+shards_per_zone := max(1, floor(num_shards / len(zones)))
+prom_zone_idx := shard_index % shards_per_zone
+    
+if zone_label == zones[prom_zone_idx] {
+    
+    assignment_idx := floor(shard_index / shards_per_zone)
+
+    if hash(address) % shards_per_zone == assignment_idx {
+        do_scrape()
+    }
+}
+```
+
+By using modulo to calculate the `prom_zone_idx`, instances will be distributed
+to zones in the sense of `A,B,C,A,B,C` and so on. This allows a modification of
+the `num_shards` value without redistribution of shards or data.
+This was preferred over allowing the number of `zones` to change, as this is
+less likely to happen.
+
+### Edge cases
+
+We have introduced asserts in the above section to warn about edge cases that
+might lead to duplicate data or data loss.
+
+By the above algorithm. prometheus instances will be distributed in an
+alternating fashion by using the already existing shard index.
+This leads to the following edge cases:
+
+When `num_shards` is 10 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 3. This yields the following distribution
+
+```
+0 1 2 | 3 4 5 | 6 7 8 | 9 | shard index
+A B C | A B C | A B C | A | zone
+0 0 0 | 1 1 1 | 2 2 2 | 0 | assignment index
+```
+
+In this case the 2nd assert will warn about double scraping of instances in
+zone A, as the same targets are being assigned to instance 0 and 9.
+
+When `num_shards` is 2 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 1. This yields the following distribution
+
+```
+0 1   | shard index
+A B C | zone
+0 0 0 | assignment index
+```
+
+In this case the 1st assert will warn about zone C not being scraped.
```

Comment:
should it be prevented by the operator? E.g. it could fail the reconciliation if shards and zones don't match.

Reply:
I don’t know what mechanisms exist in the operator.
Both could be an error stopping the reconciliation loop. At minimum the first one, because data is lost in that case. In the second case there is duplicate data, which is not ideal, but at least nothing is lost.